### PR TITLE
feat: Webhook

### PR DESCRIPTION
**Pull request by Mryashbhardwaj**
For Data Pipelines there is often a requirement that systems external to Optimus expect notification/nudge upon job state change. This helps people trigger external routines.

HTTP webhooks are supported.

**On Events:**
- Job Fail
- Job Success
- SLA Miss

**Metrics Tracked :**
- webhook request failures count 
- webhook request success count

### CODE REVIEW

File: ext/notify/webhook/webhook.go [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+	go func() {
+		s.eventChan <- event{
+			url:     attr.Route,
+			meta:    attr.JobEvent,
+			jobMeta: attr.Meta,
+			headers: attr.Headers,
+		}
+	}()
+
+	webhookQueueCounter.Inc()
```
Comment (sbchaos) [inline ext/notify/webhook/webhook.go:79]: counter will increment before the event is sent on the channel. This is sync, sending is async.

File: ext/notify/webhook/webhook_test.go [hunk 2]
```diff
@@ -0,0 +1,136 @@
+package webhook // nolint: testpackage
```
Comment (sbchaos) [inline ext/notify/webhook/webhook_test.go:1]: Can we remove this nolint, have proper test namespace.

File: ext/notify/webhook/webhook.go [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+	return nil
+}
+
+func (s *Notifier) Worker(ctx context.Context) {
+	defer s.wg.Done()
+	for {
+		select {
+		case event := <-s.eventChan:
+			client := &http.Client{}
+			payload := webhookPayload{
```
Comment (sbchaos) [inline ext/notify/webhook/webhook.go:89]: Sending http call can be extracted in another function.

**Comment by sbchaos** [comment]
Can we add documentation in this pr or next for this feature.

File: ext/notify/webhook/webhook.go [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+				Status:      event.meta.Status.String(),
+				JobLabel:    event.jobMeta.Labels,
+			}
+			payloadJSON, err := json.Marshal(payload)
+			if err != nil {
+				s.workerErrChan <- fmt.Errorf("webhook worker: %w", err)
+				continue
+			}
+
+			req, err := http.NewRequestWithContext(ctx, http.MethodPost, event.url, bytes.NewBuffer(payloadJSON))
```
Comment (sbchaos) [inline ext/notify/webhook/webhook.go:104]: We are sending ctx, which is the context used for the worker. We should have a separate context with timeout for http call.

File: ext/notify/webhook/webhook.go [hunk 5]
```diff
@@ -0,0 +1,161 @@
...
+			}
+
+			res, err := client.Do(req)
+			if err != nil {
+				s.workerErrChan <- fmt.Errorf("webhook worker: %w", err)
+				continue
+			}
+			webhookWorkerSendCounter.Inc()
+
+			err = res.Body.Close()
```
Comment (sbchaos) [inline ext/notify/webhook/webhook.go:121]: We are not checking if the call was success. Do we need to retry on failure ?

Comment (Mryashbhardwaj) [reply ext/notify/webhook/webhook.go:121]: no we do not need retries

File: ext/notify/webhook/webhook_test.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (Mryashbhardwaj) [reply ext/notify/webhook/webhook_test.go:1]: cool

File: core/scheduler/service/notification_service.go [hunk 6]
```diff
@@ -24,8 +25,15 @@ type Notifier interface {
...
 }
 
+type Webhook interface {
+	io.Closer
+	Notify(ctx context.Context, attr scheduler.WebhookAttrs) error
+}
+
 type NotifyService struct {
 	notifyChannels map[string]Notifier
+	webhookChannel Webhook
```
[the remark this answers is not in the document]
Comment (Mryashbhardwaj) [reply core/scheduler/service/notification_service.go:35]: not required

File: core/scheduler/service/notification_service.go [hunk 7]
```diff
@@ -38,9 +46,52 @@ func (n *NotifyService) Push(ctx context.Context, event *scheduler.Event) error
...
+					continue
+				}
+				secretMap = tenant.PlainTextSecrets(plainTextSecretsList).ToSecretMap()
+			}
+
+			jobWithDetails, err := n.jobRepo.GetJobDetails(ctx, event.Tenant.ProjectName(), jobDetails.Name)
+			if err != nil {
+				return err
+			}
+			headerContext := compiler.PrepareContext(compiler.From(secretMap).WithName(contextSecret))
```
[the remark this answers is not in the document]
Comment (Mryashbhardwaj) [reply core/scheduler/service/notification_service.go:69]: skipped for now

File: ext/notify/webhook/webhook.go (hunk 3 above)
Comment (Mryashbhardwaj) [reply ext/notify/webhook/webhook.go:89]: 👍

File: ext/notify/webhook/webhook.go (hunk 4 above)
Comment (Mryashbhardwaj) [reply ext/notify/webhook/webhook.go:104]: 👍

File: core/scheduler/handler/v1beta1/job_run.go [hunk 8]
```diff
@@ -24,6 +24,7 @@ type JobRunService interface {
 
 type Notifier interface {
 	Push(ctx context.Context, event *scheduler.Event) error
+	Webhook(ctx context.Context, event *scheduler.Event) error
```
[the remark this answers is not in the document]
Comment (Mryashbhardwaj) [reply core/scheduler/handler/v1beta1/job_run.go:27]: Not Required

**Review by sbchaos — APPROVED** [review]

**Pull request merged**